docs: update rebasing documentation (#9021)
renovatebot/renovate: docs/usage/updating-rebasing.md

---
title: Updating and Rebasing branches
description: How Renovate Updates and Rebases branches
---

# Updating and rebasing branches

There are many situations in which Renovate must update/rebase a branch.

Here is a list of the most common cases where Renovate must update/rebase the branch:

- When a pull request has conflicts due to changes on the base branch
- When you have enabled "Require branches to be up to date before merging" on GitHub
- When you have manually told Renovate to rebase when behind the base branch with `"rebaseWhen": "behind-base-branch"`
- When a newer version of the dependency is released

> - When you request a manual rebase from the Renovate bot

Reviewer: ```suggestion
- When you use `automerge` on a branch / pr
```

Instead, Renovate reapplies all updates into a new commit based off of the head of the base branch.

## No rebasing if you have made edits

First of all, here is the one time when Renovate _won't_ update branches.
If you edit a Renovate branch directly (e.g. to make a code fix to allow tests to pass again) then Renovate stops all updates of that branch.
It is up to you to either finish the job and merge the PR, or rename it and close it so that Renovate can take back over again.

## Rebasing conflicted PRs

If new commits to the base branch - such as merging another Renovate PR - result in an open Renovate PR having merge conflicts, then Renovate will recreate ("rebase") any conflicted PRs.
This applies both to commits to dependency files such as `package.json` as well as lock files such as `yarn.lock`.
You should not ever need to resolve such conflicts manually.
You can disable this functionality by configuring `"rebaseWhen": "never"` (not recommended).

## Rebasing out-of-date branches

There are two cases where Renovate will rebase its branches off the base branch every time they are out of date:

1. If you configure `"rebaseWhen": "behind-base-branch"`

> 1. If you use the default configuration `"rebaseWhen": "auto"` and the repository has a requirement that branches must be up-to-date before merging (e.g. "Require branches to be up to date before merging" on GitHub, or fast-forward-only settings on Bitbucket Server or GitLab)

Reviewer: ```suggestion
1. If you use the default configuration `"rebaseWhen": "auto"` and the repository has a requirement that branches must be up-to-date before merging (e.g. "Require branches to be up to date before merging" on GitHub, or fast-forward-only settings on Bitbucket Server or GitLab)
1. If you configure `automerge=true` ...
```

🙃

> 1. If you use the default configuration `"rebaseWhen": "auto"` and the repository has a requirement that branches must be up-to-date before merging (e.g. "Require branches to be up to date before merging" on GitHub, or FF-only settings on Bitbucket Server or GitLab)

Author: ```suggestion
1. If you use the default configuration `"rebaseWhen": "auto"` and the repository has a requirement that branches must be up-to-date before merging (e.g. "Require branches to be up to date before merging" on GitHub, or fast-forward-only settings on Bitbucket Server or GitLab)
```

How about this? I think `fast-forward-only` is clearer than a somewhat cryptic `FF-only`.

---
title: Updating and Rebasing branches
description: How Renovate Updates and Rebases branches
---

# Updating and rebasing branches

There are many situations in which Renovate must update/rebase a branch.

Here is a list of the most common cases where Renovate must update/rebase the branch:

- When a pull request has conflicts due to changes on the base branch
- When you have enabled "Require branches to be up to date before merging" on GitHub
- When you have manually told Renovate to rebase when behind the base branch with `"rebaseWhen": "behind-base-branch"`
- When a newer version of the dependency is released
- When you request a manual rebase from the Renovate bot

> - When you use `automerge` on a branch / pr

Reviewer: ```suggestion
- When you use `"automerge": true` and `"rebaseWhen": "auto"` on a branch / pr
```

Is this clear enough ?

Renovate uses its own version of "rebasing", which is _not the same_ as doing a `git rebase` with Git.
Instead, Renovate reapplies all updates into a new commit based off of the head of the base branch.

## No rebasing if you have made edits

First of all, here is the one time when Renovate _won't_ update branches.
If you edit a Renovate branch directly (e.g. to make a code fix to allow tests to pass again) then Renovate stops all updates of that branch.
It is up to you to either finish the job and merge the PR, or rename it and close it so that Renovate can take back over again.

## Rebasing conflicted PRs

If new commits to the base branch - such as merging another Renovate PR - result in an open Renovate PR having merge conflicts, then Renovate will recreate ("rebase") any conflicted PRs.
This applies both to commits to dependency files such as `package.json` as well as lock files such as `yarn.lock`.
You should not ever need to resolve such conflicts manually.
You can disable this functionality by configuring `"rebaseWhen": "never"` (not recommended).

## Rebasing out-of-date branches

> There are two cases where Renovate will rebase its branches off the base branch every time they are out of date:

Author: ```suggestion
There are multiple cases where Renovate will rebase its branches off the base branch every time they are out of date:
```

1. If you configure `"rebaseWhen": "behind-base-branch"`
1. If you use the default configuration `"rebaseWhen": "auto"` and the repository has a requirement that branches must be up-to-date before merging (e.g. "Require branches to be up to date before merging" on GitHub, or fast-forward-only settings on Bitbucket Server or GitLab)

> 1. If you configure `automerge=true` ...

Author: L41 is still a bit vague right now:

```
1. If you configure `automerge=true` ...
```

Do we want to explain more here, or do we just want to that using `automerge=true` is one of the cases where Renovate will rebase its branches off the base branch.

Reviewer: yes, if branch is out of date and rebaseWhen is auto -> renovate will rebase. was just a starter to complete by you 😉

Author: Can you take a look at the new L41 and tell me what you think?
